File: backend/app/models/habit_log.py
from datetime import datetime, date, timezone
from sqlalchemy import Integer, String, DateTime, Boolean, ForeignKey, Date
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.db.database import Base


class HabitLog(Base):
    __tablename__ = "habit_logs"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    habit_id: Mapped[int] = mapped_column(Integer, ForeignKey("habits.id"), nullable=False, index=True)
    date: Mapped[date] = mapped_column(Date, nullable=False, index=True)
    completed: Mapped[bool] = mapped_column(Boolean, default=False)
    completed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    skipped_reason: Mapped[str | None] = mapped_column(String(300), nullable=True)
    note: Mapped[str | None] = mapped_column(String(500), nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc)
    )

    # Relationships
    habit = relationship("Habit", back_populates="logs")

File: mobile/lib/models/habit_log.dart
class HabitLog {
  final int id;
  final int habitId;
  final DateTime date;
  final bool completed;
  final DateTime? completedAt;
  final String? note;
  final String? skippedReason;
  final DateTime createdAt;

  HabitLog({
    required this.id,
    required this.habitId,
    required this.date,
    required this.completed,
    this.completedAt,
    this.note,
    this.skippedReason,
    required this.createdAt,
  });

  factory HabitLog.fromJson(Map<String, dynamic> json) {
    return HabitLog(
      id: json['id'],
      habitId: json['habit_id'],
      date: DateTime.parse(json['date']),
      completed: json['completed'] ?? false,
      completedAt: json['completed_at'] != null
          ? DateTime.parse(json['completed_at'])
          : null,
      note: json['note'],
      skippedReason: json['skipped_reason'],
      createdAt: DateTime.parse(json['created_at']),
    );
  }
}

